lib/chef/provider/package/macospkg.rb

```diff
@@ -0,0 +1,80 @@
+#
+# Author:: Joshua Timberman (<jtimberman@chef.io>)
+# Author:: William Theaker (<william.theaker+chef@gusto.com>)
+# Copyright:: Copyright (c) Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../provider/package"
+require_relative "package"
+
+class Chef
+  class Resource
+    class MacosPkg < Chef::Provider::Package
+      provides:macos_pkg
+
+      def load_current_resource
+        @current_resource = Chef::Resource::Package.new(new_resource.name)
+        current_resource.package_name(new_resource.package_name)
+        current_resource.version(get_current_version)
+        logger.trace("#{new_resource} current package version: #{current_resource.version}") if current_resource.version
+
+        download_pkg if new_resource.source
+
+        current_resource
+      end
+
+      def get_current_version
+        shell_out("pkgutil --pkg-info '#{new_resource.package_id}'").stdout.to_s[/version: (.*)/, 1]
+      end
+
+      def define_resource_requirements
+        requirements.assert(:install) do |a|
+          a.assertion { new_resource.source || new_resource.file }
+          a.failure_message Chef::Exceptions::Package, "Must provide either a file or source property for #{new_resource.package_name} macos_pkg resource."
+        end
+      end
+
+      def install_package(name, version)
+        shell_out("installer -pkg #{pkg_file} -target #{new_resource.target}")
+      end
+
+      def upgrade_package(name, version)
+        shell_out("pkgutil --forget '#{new_resource.package_id}'") if current_resource.version
```

Comment:
I don't think this removes the files. We probably want a ruby version of:
https://gist.github.com/githubutilities/b5318d08a4b970d104f1

Otherwise files A, B, C come with version 1, then versions, A, B, D, come with version 2 (`C` is now abandoned), then files A, D, E, F come with version 3 (`B` and `C` are now both abandoned). Not only is that  a bit ugly, but it can actually cause real issues, lots of software does things like load-all-files-in-directory.

So we should make an `uninstall_package`, and then `upgrade` should call `uninstall` and then `install`

Reply:
>> `pkgutil --forget`
> I don't think this removes the files.

It doesn't, but we need to forget the package receipt to allow downgrading packages or upgrading improperly versioned packages.

In general, we shouldn't treat the bill of materials/BOM file as the canonical list of files installed by a package, since it's just a snapshot of the package's `Payload` directory tree that's copied into a filesystem.

To use a simplified version of the `MXFPlugIns.pkg` package in Apple's [Pro Video Formats 2.3](https://support.apple.com/kb/DL2100) as an example:
```
.
├── Bom
├── PackageInfo
├── Payload
│   └── Library
│       └── Video
│           └── foo
└── Scripts
│   └── postinstall_actions
│       └── moveMXFSettings.sh
│   ├── postinstall
│   └── preinstall
```

The script in `Scripts/postinstall_actions/moveMXFSettings.sh` moves files from one directory into one not defined in the BOM, which means they get left behind when iterating over the package's BOM to delete files. Renaming or moving files is a very common pattern you'll see across [different packages](https://github.com/sheagcraig/yo/blob/master/pkg/postinstall#L6), so it's not just an Apple deployment quirk.

On the point of cleaning up conflicting files when upgrading, we should rely on package maintainers to build [preinstall](https://github.com/sheagcraig/yo/blob/master/pkg/preinstall) scripts to do that instead, in part because removing the binary of a running process can cause crashes.

Another caveat, this implementation also leaves behind empty directories since we can't be sure the directories are exclusively owned by a package. For example, an organization might deliver a data bag secret via a separate package, but we wouldn't want to wipe out `/etc/cinc` when upgrading between minor versions of Chef/Cinc.

For what it's worth, I wouldn't use the remove action myself and would instead rely on either a [vendor's uninstall method](https://community.carbonblack.com/t5/Knowledge-Base/Carbon-Black-Cloud-How-To-Uninstall-Apple-macOS-Sensor-v3-x-x-x/ta-p/65417) or write a custom recipe to delete the package and its artifacts. I think the lack of native uninstall methods in macOS packages might also be why [Puppet](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/provider/package/apple.rb#L4) only supports installing packages on Darwin, but not removing them.

Reply:
@jaymzh Do the above comments resolve your concerns?

Reply:
It sounds like it would be useful to create the hooks as @jaymzh mentioned, and then the cookbook author would be responsible for filling those out in the case there are blocking processes, etc... to ensure that the binaries could cleanly be removed. But {uninstall is exercise left up to the cookbook writer} and not something that we generically create a pattern for, just the hooks to invoke those actions.

Reply:
I'm with Thomas:

I would, at the VERY LEAST make a property that applies to `:upgrade` that at least does it's best to clean up. If people don't want it they can turn it off. And yeah we shouldn't delete directories if anything is left in them, just like RPM and friends dont.

Once that code is written it seems useful to offer an `:uninstall` based on it. It's not perfect, but there's literally no better way. IT'd do the `forget` and the cleanup, which is better than what anyone else offers.